File: tennis-game/kotlin-arrow-tennis-game/src/main/kotlin/GamePlay.kt
import GameInteraction.readPlayer
import GameInteraction.showGameScore
import arrow.fx.IO
import arrow.fx.extensions.fx
import arrow.optics.Lens

object GamePlay {

    fun playTennisGame(game: Game): IO<Game> =
            IO.fx {
                val playerScoring = !readPlayer()
                val updatedGame = !trackScoredPoint(playerScoring, game)
                val shownGameScore = !showGameScore(updatedGame)
                when {
                    gameCompleted(shownGameScore) -> shownGameScore
                    else -> !playTennisGame(shownGameScore)
                }
            }

    private val trackScoredPoint: (ScoringPlayer, Game) -> IO<Game> = { scoringPlayer, game ->
        IO.fx {
            trackScorePointForPlayer(scoringPlayer, game)
        }
    }

    val trackScorePointForPlayer: (ScoringPlayer, Game) -> Game = { scoringPlayer, game ->
        when (scoringPlayer) {
            Player1 -> update(game, gameToPlayer1Score, gameToPlayer2Score)
            Player2 -> update(game, gameToPlayer2Score, gameToPlayer1Score)
        }
    }

    private val gameCompleted: (Game) -> Boolean = { game ->
        gameToPlayer1Score.get(game) == Wins || gameToPlayer2Score.get(game) == Wins
    }
    private val update: (
            game: Game,
            scoringPlayerScore: Lens<Game, Score>,
            opponentPlayerScore: Lens<Game, Score>
    ) -> Game = { game, scoringPlayerScore, opponentPlayerScore ->
        when (scoringPlayerScore.get(game)) {
            Love -> scoringPlayerScore.set(game, Fifteen)
            Fifteen -> scoringPlayerScore.set(game, Thirty)
            Thirty -> scoringPlayerScore.set(game, Forty)
            Forty -> when (opponentPlayerScore.get(game)) {
                Advantage -> opponentPlayerScore.set(game, Forty)
                Forty -> scoringPlayerScore.set(game, Advantage)
                else -> scoringPlayerScore.set(game, Wins)
            }
            Advantage -> scoringPlayerScore.set(game, Wins)
            Wins -> game
        }
    }
}

File: tennis-game/typescript-fpts-tennis-game/src/GamePlay.ts
import {
    advantage,
    fifteen,
    forty,
    Game,
    gameToPlayer1Score,
    gameToPlayer2Score,
    Score,
    ScoringPlayer,
    thirty,
    wins
} from "./Data";
import {Lens} from "monocle-ts";
import {of, Task} from "fp-ts/Task";

export const trackScoredPoint: (scoringPlayer: ScoringPlayer, game: Game) => Task<Game> =
    (scoringPlayer: ScoringPlayer, game: Game) => of(trackScorePointForPlayer(scoringPlayer, game))

export const gameCompleted: (game: Game) => boolean = (game: Game) =>
    gameToPlayer1Score.get(game)._tag === 'Wins' || gameToPlayer2Score.get(game)._tag === 'Wins'

const trackScorePointForPlayer: (scoringPlayer: ScoringPlayer, game: Game) => Game =
    (scoringPlayer: ScoringPlayer, game: Game) => {
        switch (scoringPlayer._tag) {
            case 'Player1':
                return update(game, gameToPlayer1Score, gameToPlayer2Score)
            case 'Player2':
                return update(game, gameToPlayer2Score, gameToPlayer1Score)
        }
    }

const update: (game: Game, scoringPlayerScore: Lens<Game, Score>, opponentPlayerScore: Lens<Game, Score>) => Game =
    (game: Game, scoringPlayerScore: Lens<Game, Score>, opponentPlayerScore: Lens<Game, Score>) => {
        switch (scoringPlayerScore.get(game)._tag) {
            case 'Love':
                return scoringPlayerScore.set(fifteen())(game)
            case 'Fifteen':
                return scoringPlayerScore.set(thirty())(game)
            case 'Thirty':
                return scoringPlayerScore.set(forty())(game)
            case 'Forty': {
                switch (opponentPlayerScore.get(game)._tag) {
                    case 'Advantage':
                        return opponentPlayerScore.set(forty())(game)
                    case 'Forty':
                        return scoringPlayerScore.set(advantage())(game)
                    default:
                        return scoringPlayerScore.set(wins())(game)
                }
            }
            case 'Advantage':
                return scoringPlayerScore.set(wins())(game)
            case 'Wins':
                return game
        }
    }
